Implement deep merging of fields

ycombinator opened the pull request:
Resolves #3628.

Consider the following Filebeat configuration:

```yml
fields_under_root: true
fields:
  my.field1: "field1"
  my.field2: "field2"

- type: stdin
  enabled: true

  fields_under_root: true
  fields:
    my.field3: "field3"
```

Prior to this PR, Filebeat would not merge general fields and per-input fields deeply. That is, the event resulting from the above configuration would look like:

```json
{
   "my": {
      "field3": "field3"
    }
}
```

With this PR, the resulting event looks like:
```json
{
   "my": {
      "field1": "field1",
      "field2": "field2",
      "field3": "field3"
    }
}
```

ycombinator commented:
For reviewers:

1. I left the original `common.MergeFields` as-is since it's an exported function. I created a new exported function `common.MergeFieldsDeep` and extracted common code from both into an un-exported helper function. WDYT?

2. I have not yet added a CHANGELOG entry for this PR because I want to know: should this be considered a bug fix or a breaking change to behavior of FIlebeat w.r.t fields or both?

urso commented on libbeat/common/mapstr.go:
```diff
@@ -277,28 +277,57 @@ func MergeFields(ms, fields MapStr, underRoot bool) error {
+//
+// An error is returned if underRoot is true and the value of ms["fields"] is not a
+// MapStr.
+func MergeFieldsDeep(ms, fields MapStr, underRoot bool) error {
```
Yay, more methods/functions operating on MapStr :(

Let's rename 'ms' to 'target' and 'fields' to 'from', to make it more clear that we do overwrite fields in the original MapStr.

Note: I read a 'merge' function as something that does not destructively modifies one of its inputs.

ycombinator replied:
> Note: I read a 'merge' function as something that does not destructively modifies one of its inputs.

I was keeping the naming and behavior consistent with the older `MergeFields` function, to try and indicate that `MergeFieldsDeep` behaves just like `MergeFields` (it modifies one of the inputs) but in a recursive/deep fashion.

I'm happy to change the name to something else or make the function return a new `MapStr` instead of modifying an input. Or perhaps we should make this a method on `MapStr`? To me, it doesn't seem generic enough to become a method (since it does some very specific stuff with the `fields` field) which is why I stayed away from that option.

urso replied:
I'm ok keeping the name `MergeFieldsDeep`.

urso commented:
Please double check the processors in libbeat/processors, if they need to be adapted as well (e.g. add_fields processor).

> I have not yet added a CHANGELOG entry and some labels for this PR because I want to know: should this be considered a bug fix or a breaking change to behavior of FIlebeat w.r.t fields or both?

For me it is clearly a bug fix. Fields were always supposed to be merged, with global fields potentially overwriting local ones (overwriting at leaf-level, not top-level).

ycombinator commented:
> Please double check the processors in libbeat/processors, if they need to be adapted as well (e.g. add_fields processor).

* [x] `add_cloud_metadata`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `add_docker_metadata`: using the `DeepUpdate` method on `MapStr`
* [x] `add_fields`: using the `DeepUpdate` method on `MapStr`
* [x] `add_host_metadata`: using the `DeepUpdate` method on `MapStr`
* [x] `add_kubernetes_metadata`: using the `DeepUpdate` method on `MapStr`
* [x] `add_labels`: using the `DeepUpdate` method on `MapStr`
* [x] `add_locale`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `add_observer_metadata`: using the `DeepUpdate` method on `MapStr`
* [x] `add_process_metadata`: using the `Put` method on `MapStr`
* [x] `add_tags`: for adding tags to existing field, appends tags if it is an array field; else throws error
* [x] `community_id`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `convert`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `copy_fields`: using the `Put` method on `MapStr`
* [x] `decode_base64_field`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `decode_cef` (X-Pack): using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `decode_csv_fields`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `decode_json_fields`: **doing a simple event.Fields[k] = v, which could lead to overwriting of fields!**
* [x] `decompress_gzip_field`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `dissect`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `dns`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `drop_event`: Doesn't add new fields to event
* [x] `drop_fields`: Doesn't add new fields to event
* [x] `extract_array`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `include_fields`: using the `Put` method on `MapStr`
* [x] `registered_domain`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `rename`: using the `Put` method on `MapStr`
* [x] `script`: does arbitrary processing of event according to user-specified script
* [x] `timestamp`: using the `Put` method on `MapStr` (called via `event.PutValue`)
* [x] `truncate_fields`: using the `Put` method on `MapStr` (called via `event.PutValue`)

So only the `decode_json_fields` processor needs to be fixed IMO. Are you okay with me doing this in a separate PR?

ycombinator commented:
Hi @urso, thanks for the review. I've addressed all your feedback and this is ready for your review again, when you get a chance.

In particular, please see my comment/question about naming the function [in this comment](https://github.com/elastic/beats/pull/13909#discussion_r331477180) and my question about a follow up PR for fixing the field setting in the `decode_json_fields` processor in [this comment](https://github.com/elastic/beats/pull/13909#issuecomment-538447772).

Thanks!

urso commented:
> So only the decode_json_fields processor needs to be fixed IMO. Are you okay with me doing this in a separate PR?

Sure.

ycombinator commented:
jenkins, test this

ycombinator commented:
>> So only the `decode_json_fields` processor needs to be fixed IMO. Are you okay with me doing this in a separate PR?

> Sure.

Created issue to track this work: https://github.com/elastic/beats/issues/13939
